README.md
# Swami - README
### Copyright (C) 1999-2019 Element Green and others
http://www.swamiproject.org

##### NOTE that this readme might be outdated.


## 1. What is Swami?

Swami (Sampled Waveforms And Musical Instruments) is a SoundFont
editor. SoundFont files are a collection of audio samples and
other data that describe instruments for the purpose of composing
music. SoundFont files do not describe the music itself, but rather the
sounds of the instruments. These instruments can be composed of any
digitally recordable or generated sound. This format provides a
portable and flexible sound synthesis environment that can be
supported in hardware or software.

## 2. What happend to the Smurf SoundFont Editor?

Nothing actually. Since I was already doing an entire re-code of Smurf
I decided it would be good to change the name to give the program a
fresh start and because I'm planning to make it more than just a
SoundFont editor. Also the word 'Smurf' is copyrighted so I thought a
change wise in this day and age of law suits. Swami does not yet
completely replace Smurf in functionality, but it will.

## 3. License

See [Copying](https://github.com/swami/swami/blob/master/COPYING).

## 4. Changes from Smurf

Most of the changes have occured in the programming architecture. Much
of this work has been done in preparation for really cool features
(TM) :) The SoundFont editing code is now completely abstracted from
the GUI in separate shared libraries making things like: scripting
support, multiple client concurrent access to SoundFont objects and SwamiJam
(networked semi-realtime composition with friends) more feasible. Most
importantly things are much cleaner now so I don't go insane in attempts
to add functionality.

New features:
- FluidSynth soft synth support (actually its required) which allows any
  OS supported sound card to be used and gives us modulators
  (real time effects) and routeable digital audio output for adding additional
  effects and for easy recording.
- More operations work for multiple items. Including loading/saving/closing
  of multiple SoundFont files, loading multiple samples, and
  setting/viewing properties on multiple items.
- Modulator editor for editing real time control parameters
- Default toggle buttons for generator (effect) controls

Things missing (as of this README):
- Undo support
- virtual SoundFont bank support
- OSS AWE/SB Live! wavetable backend


## 5. Requirements

Look at the INSTALL file for instructions on compiling and installing Swami
and for more details on software requirements.

Version numbers listed are what I developed Swami on, older or newer versions
of the listed software may or may not work, but give it a try! And do let me
know if you encounter any problems.

Swami has the following requirements:
- Linux, Windows and Mac OS X.  Probably anywhere else GTK+ will run.
- GTK+ v2.0
- FluidSynth v2.0 (software wavetable synthesizer)
- libsndfile

GTK homepage: http://www.gtk.org
FluidSynth homepage: http://www.fluidsynth.org
libsndfile homepage: http://www.mega-nerd.com/libsndfile
ALSA homepage: http://www.alsa-project.org
OSS homepage: http://www.4front-tech.com


## 6. Supported sound cards

Any FluidSynth supported sound card (i.e. works under ALSA or OSS in Linux), since
Swami uses FluidSynth which does all synthesis in software.  A wavetable
backend is planned for using hardware based SoundFont synthesizers like the
SB AWE/Live! for those who don't want to use their precious CPU cycles :)


## 7. Features

Current features include:
- Loading of multiple SoundFont 2.x files *.SF2
- Saving of multiple SoundFont files
- Copy Samples/Instruments/Presets between and within SoundFont objects
- Load and export multiple audio sample files (WAV, AU, etc)
- View/set multiple Preset, Instrument and Sample parameters
- A GTK piano widget to play notes with computer keyboard/mouse
- Wavetable loading of entire SoundFont
- Piano MIDI controls (Pitch bender amount, volume etc.)

Features provided by FluidSynth:
- Flexible driver system that supports OSS, ALSA, LADSPA, Jack, Midishare, etc
- Real time effects via modulators and C API

## 8. Future plans

The following are in the works or planned:
- Undo support (just haven't written a front end for it really)
- Copy/Paste generator parameters
- Graphical manipulation of generator parameters

Was part of Smurf, not yet added:
- Loading, saving and editing of virtual bank (*.bnk) files
- Multiple undo/redo tree system, allowing for multiple redo "branches"
- Sample cut support
- Sample waveform generation
- International language support (currently disabled until there are some
  updated translations, contact me if interested)


## 9. Troubleshooting drivers (or why does Swami crash on startup?)

[Some of these switches haven't been enabled yet!]

Some new command line switches were added to help with trouble shooting
problems with Swami. Crashes experienced on startup of Swami are often caused
by audio drivers.  Type "swami --help" to get a list of usable switches.
Since Swami, by default, tries to open the drivers on start up, a way to
disable this behavior was added. "swami -d" will cause Swami to start, but
drivers will not automatically be opened. Once started you can change your
driver preferences (set certain ones to NONE) to try to determine which
driver is crashing. Then issue a Control->"Restart Drivers" to cause the
drivers to be re-opened.

Swami v0.9.0pre1
Copyright 1999-2002 Josh Green <jgreen@users.sourceforge.net>
Distributed under the GNU General Public License
Usage: swami [options]
Options:
-h, --help       Display this information
-d               Disable drivers (try if Swami crashes on start up)
-c               Ignore preferences and use factory defaults


## 10. Using Virtual Keyboard interface

[Most of this is true, although some things haven't been implemented yet]

The wavetable and piano interface have been changed in Smurf v0.52.  Some of
the new features include:
* MIDI channel control (on toolbar, titled "Chan")
	Sets what MIDI channel the piano plays on. All other MIDI controls
	(bank, preset and MIDI controls on pop up menu) are remembered for
	each channel.
* MIDI bank and preset controls (also on toolbar)
	Sets the current bank and preset for the selected MIDI channel. These
	are often changed automatically by the user interface when selecting
	items in the SoundFont tree, they can be changed manually as well.
* Entire SoundFont loading
	Allows you to load an entire SoundFont. This is particularly useful
        to make a SoundFont available for sequencing with other programs.
* Controls for changing how interface works with wavetable
	* Piano follows selection
	
		When enabled: virtual keyboard will "follow" the SoundFont
		tree selection. This will cause the selected item to be the one
		heard on the piano. If selected item is a Preset in a loaded
		SoundFont, then the piano will select the Bank:Preset of that
		item. Otherwise it is set to the Temporary Audible Bank:Preset
		(configurable from preferences, defaults to 127:127). By
		disabling this feature you can have more control over what is
		being played by the virtual keyboard.

	* Auto load temp audible
	
		When enabled: selected items in the SoundFont tree that aren't
		Presets in a loaded SoundFont, are automatically loaded into
		the wavetable device. They are mapped to a temporary
		Bank:Preset which is configurable from Preferences. Disabling
		this allows for faster editing, if one does not care to hear
		items.

* Temporary audible -
The temporary audible enables you to listen to components of a SoundFont that
aren't directly mapped to a MIDI Bank:Preset. This is done by designating a
Bank:Preset for temporary use. It defaults to 127:127 but can be changed
in Preferences. When a non-Preset or a Preset that isn't in a loaded SoundFont
is loaded (either from the "Auto load temp audible" option or manually from the
"Wavetable Load" item on the right click menus), it is mapped to this temporary
bank and preset number.


## 11. Creating Preset/Instrument zones

To add zones to an Instrument hold down the CTRL key and select the samples
you want to be added (SHIFT works too for selecting a range of items). Then
"Right Click" the Instrument you want to add the samples to.
Select "Paste" from the pop up menu. A zone will be created for each
Sample (or Instrument zone) you selected. This same procedure can be applied to
Presets using Instruments as zones.
To create a "Global Zone", "Right Click" on the Preset or Instrument you want
to create the zone under. Select "Global Zone" from the pop up menu. Global
Zones are used to set "Default" generator (and modulator) parameters that the
other zones don't already have set.


## 12. Using the sample viewer

The sample viewer is used to change loop points for samples and
instrument zones.

Some things have changed in the sample viewer and probably will be
changing again. The sample zoom/unzoom is now performed with the SHIFT
key held down.  Regular left mouse clicks are now used for sample
selections for audio editing functions (currently only "cut").

Tips to using the sample viewer:

Zooming IN/OUT
Hold down SHIFT on the computer keyboard and click and hold the LEFT mouse
button in the sample window, a white line will appear to mark your click, move
the mouse to the left or the right to zoom into the sample. The distance you
move the mouse from the original point of your click determines the speed in
which the view zooms. If you then move the mouse to the other side of the line
the view will unzoom. The RIGHT mouse button does the same thing, except the
first direction you move the mouse unzooms the view.

Selecting audio data
Left click and drag to mark an area. An existing selection can be
changed by holding down CTRL and Left or Right clicking to change the
left or right ends of the selection respectively.


## 13. Undo system

**Not enabled yet!**

Swami has a rather flexible undo system. Many actions aren't currently
undo-able, but will be implimented soon. Currently all actions
relating to new SoundFont items, pasted SoundFont items, and deleted
items are undo-able.  The undo system is based on a tree rather than
queue or list. This means that multiple changes can be easily tested.

A traditional undo system is oriented as a list. Here is an example to help
understand the Swami undo tree system compared with the traditional one.

1. You change some parameters related to the Modulation Envelope of an
   instrument
2. You don't quite like the changes, so you undo them
3. You try some other values for the same Modulation Envelope

Lets say now you want to try the parameters in Step 1 again to compare to the
settings in Step 3. In a traditional undo system you would not be able to do
this because Step 3 would cause the "redo" information from Step 2 to be lost.
Using a tree, multiple branches of "state" information can be stored. So rather
than Step 3 "over writing" the redo info from Step 2, it would simply create
another "branch" in the tree.

Note: This was just an example of usage, but Swami doesn't support undo-ing of
parameter changes yet.

When just using the Undo/Redo menu items the undo system acts as a
traditional list oriented undo system.
The "Undo History" item on the Edit menu will bring up a dialog that
lists previous actions. The dialog buttons available are Undo, Redo,
Jump, Back, Forward, < (left arrow), > (right arrow).
The current location in the state history is indicated by 2 greater
than symbols ">>" in the left most column.

Undo and Redo act just like the menu items, undo-ing (moving up the
undo history) and redo-ing (moving down the undo history) the
next/previous action from the current location.
The "Jump" button will jump to the desired position in the undo
history, causing actions to be undone/redone as necessary to realize
the desired state.
The Back and Forward buttons act like the traditional web browser
function.  Currently the last 10 positions in the undo history are
remembered. This makes testing different settings easier. This isn't
really useful yet as the currently undo-able actions aren't really
ones you would wan't to test different settings with. But in the
future when generator parameters are tracked in the undo history, this
will be useful.
The < (left arrow) and > (right arrow) are the second and third column
headers of the history list. These columns are used to indicate other
branches in the tree. Items that have other "sibling" branches will
have a number in brackets (example: [ 1]) listed in the < (left arrow)
and/or > (right arrow) column(s). The left arrow switches to older
branches the right arrow to newer ones. Numbers in these columns
indicate how many alternate older or newer branches (left arrow or
right arrow respectively) are available.

To switch to an alternate branch, select an item that has a number in
one of the arrow columns. Click the corresponding arrow button at the
top of the list.  The selected item and all items below will be
changed to reflect the new branch.


## 14. Virtual Banks

**Not implemented yet in Swami!**

Virtual banks allow you to make a SoundFont bank from many other SoundFont
files. Swami currently supports loading and saving of the AWE .bnk format.
This is a simple text file format listing each Preset and where it can be found
SoundFont file:Bank:Preset and its destination Bank:Preset.

**NOTE:** Make sure you set the "Virtual SoundFont bank search paths" field
in Preferences. It should be a colon seperated list of where to find your SoundFont
files.  Example: "/home/josh/sbks/*:/tmp/sbks". Adding a '/*' to the end of a
path will cause that directory to be recursively scanned (all sub directories
under it).

**Current limitations (will be fixed in future)**

SoundFont files will not automatically be loaded with a virtual bank, so you'll
have to load them by hand.
Default SoundFont bank can't be cleared
Can't edit virtual bank mappings


#### Adding a Preset to a virtual bank
Simply select and paste Preset items into the Mapping section of the virtual
bank.

#### Setting default SoundFont
The default SoundFont sets the default Preset instruments. The Preset mappings
override the default Presets. To change the default SoundFont simply paste a
SoundFont item to the <DEFAULT> branch of the virtual bank.


## 15. SoundFont information, links and other resources

I wrote an "Intro to SoundFonts" and its available on the Swami homepage
http://swami.sourceforge.net. Its a general overview of what a SoundFont is.

More information on the SoundFont standard can be obtained from
www.soundfont.com. There is a PDF of the SoundFont technical specification
at www.soundfont.com/documents/sfspec24.pdf.  If you find that it has moved
then look under the Resources section off of the soundfont.com home page
(although it looks like you have to get a login to go there).

Thanks to Frank Johnson at EMU for clarifying this.


## 16. Thanks

Thanks to the following projects for making Swami possible:
- FluidSynth (its what makes the instruments sound :)
- gstreamer for gobject2gtk code and helpful examples of using GObject
- Glade GTK interface builder
- Gimp for graphics editing
- Blender for render of campfire in desert Splash image
- xmms code which I borrowed the multi selection file dialog modifications from
- Piano to computer keyboard mapping inspired from soundtracker


## 17. Trademark Acknowledgement

SoundFont is a registered trademark of E-mu Systems, Inc.
Sound Blaster Live! and Sound Blaster AWE 32/64 are registered trademarks of
Creative Labs Inc.
All other trademarks are property of their respective holders.


## 18. Contact

Contact me (Element Green) particularly if you are interested in helping
develop Swami. Programming, web page design, and
documentation are all things that are lacking, so if you have an
interest, do contact me. Also contact me if you find any bugs, have
trouble compiling on a platform you think should be supported, want to
tell me how much you like Swami, or have suggestions or ideas.

Email: element@elementsofsound.org

Swami homepage: http://www.swamiproject.org

=====================================


NEWS
+-------------------------------------------------------+
+	Smurf Sound Font Editor	- News			+
+	Copyright (C) 1999-2002 Josh Green		+
+	Email: jgreen@users.sourceforge.net		+
+	Smurf homepage: http://smurf.sourceforge.net	+
+-------------------------------------------------------+

Jun 5th 2002  - Swami v0.9.0pre1 released (yee ha!)

--- Smurf Sound Font Editor (RIP) ---

Sep 6th 2001  - Smurf v0.52.5 released
May 5th 2001  - Smurf v0.52.1 released
May 2nd 2001  - Smurf v0.52 released
Jan 31st 2001 - Smurf v0.50.1 released
Jan 1st 2001  - Smurf v0.50.0 released
Sep 12th 2000 - Smurf v0.49.8 released
Jul 5th 2000  - Smurf v0.49.5 released
Mar 25th 2000 - Smurf v0.49.1 released
Mar 20th 2000 - Smurf v0.49 released
Dec 6th 1999  - Smurf v0.46 released
Nov 8th 1999  - Smurf v0.45 released
Sep 12th 1999 - Smurf v0.40 released, first public release!


patchesdb/list.py
#!/usr/bin/python
#
# list.py - List patch items.
#
# PatchesDB
# Copyright (C) 2005-2007 Josh Green <jgreen@users.sourceforge.net>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; version 2
# of the License only.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA or point your web browser to http://www.gnu.org.

import cgi
from urllib import urlencode

from PatchItem import *
import InstDB

def action (req, fields):
    Type = fields.get ("Type", "Patch")
    if Type == "Inst": s = "Instruments"
    elif Type == "Sample": s = "Samples"
    else: s = "Patch Files"

    req.write (InstDB.header ("List " + s))

    if Type == "Inst": list_insts (req, fields)
    elif Type == "Sample": list_samples (req, fields)
    else: list_patches (req, fields)

    req.write (InstDB.footer ())


def list_patches (req, fields):
    OrderBy = fields.get ("OrderBy", None)
    Category = int (fields.get ("Category", 0))
    CategoryExact = int (fields.get ("CategoryExact", 0))

    OrderDir = int (fields.get ("OrderDir", 0))
    if OrderDir < 0 or OrderDir > 2: OrderDir = 0

    Offset = int (fields.get ("Offset", 0))
    Count = int (fields.get ("Count", 50))

    if Count <= 5: Count = 5
    elif Count > InstDB.MaxListItems: Count = InstDB.MaxListItems

    itemFields = list (PatchHelper.list_format)

    # Category specified?
    if Category:
        CategoryIds = (str(Category),)
        if not CategoryExact:  # Exact category NOT specified?
            InstDB.Cur.execute ("SELECT CategoryName FROM Category"
                                " WHERE CategoryId=%s", Category) # Name
            res = InstDB.Cur.fetchone ()
            if res:   # Get list of sub categories (if any)
                InstDB.Cur.execute ("SELECT CategoryId FROM Category"
                                    " WHERE LEFT(CategoryName, %s) = %s",
                                    (len (res["CategoryName"]) + 1,
                                     res["CategoryName"] + ':'))
                for row in InstDB.Cur.fetchall():
                    CategoryIds += (str(row["CategoryId"]),)

        # Remove category from item fields
        if "CategoryName" in itemFields:
            itemFields.remove ("CategoryName")

    # Create SQL query
    sel = PatchHelper.sql_select (itemFields)

    # Make sure order by field is valid
    if OrderBy and not PatchHelper.field_searchable (OrderBy): OrderBy = None
    if not OrderBy: OrderBy = "FileName"

    sel.orderby = PatchHelper.sql_order_field (OrderBy, OrderDir)
    sel.limit = "%d,%d" % (Offset, Count)
    sel.where.append ("PatchInfo.State = 'Active'")
    sel.calc_found_rows = True

    if Category:
        # Match categories (1st, 2nd or 3rd)
        s = string.join (CategoryIds, ',')
        sel.where.insert (0, "(PatchInfo.CategoryId IN (%s)"
                          " || PatchInfo.CategoryId2 IN (%s)"
                          " || PatchInfo.CategoryId3 IN (%s))" % (s, s, s))
        if not "PatchInfo.CategoryId" in sel.fields:
            sel.fields += ("PatchInfo.CategoryId",)

    InstDB.Cur.execute (sel.query ())
    rows = InstDB.Cur.fetchall ()

    # Get total number of matching items
    InstDB.Cur.execute ("SELECT FOUND_ROWS() AS Total");
    row = InstDB.Cur.fetchone ()
    total = int (row["Total"])

    # Re URL-ify the input fields
    pager_urlstr = urlencode ((("Action", "list"), ("Type", "Patch"),
                               ("Category", Category), ("OrderBy", OrderBy),
                               ("OrderDir", OrderDir), ("Count", Count)))
    pager_urlstr = cgi.escape (pager_urlstr)

    # Display the pager bar
    InstDB.show_pager (req, pager_urlstr, Offset, Count, total)

    # Create clickable titles that sort output
    urlstr = urlencode ((("Action", "list"), ("Type", "Patch"),
                         ("Category", Category), ("Count", Count)))
    urlstr = cgi.escape (urlstr)

    sortList = []
    for field in itemFields:
        if PatchHelper.field_searchable (field):
            sortList += ((field, PatchHelper.field_title (field)),)
        else: sortList += (PatchHelper.field_title (field),)

    titles = sort_titles (sortList, urlstr, OrderBy, OrderDir, PatchHelper)

    if rows:
        table = InstDB.tabular (titles)
        table.tableattr = 'width="100%"'

        for row in rows:
            req.write (table.addrow (PatchHelper.html_vals (itemFields, row)))

        req.write (table.end ())
    else:
        req.write ("<b>No files match criteria</b>\n")

    # Display the pager bar
    InstDB.show_pager (req, pager_urlstr, Offset, Count, total)


def list_insts (req, fields):
    OrderBy = fields.get ("OrderBy", None)

    OrderDir = int (fields.get ("OrderDir", 0))
    if OrderDir < 0 or OrderDir > 2: OrderDir = 0

    Offset = int (fields.get ("Offset", 0))
    Count = int (fields.get ("Count", 50))
    PatchId = int (fields["PatchId"])

    if Count <= 5: Count = 5
    elif Count > InstDB.MaxListItems: Count = InstDB.MaxListItems

    InstDB.Cur.execute ("SELECT State, UserId FROM PatchInfo WHERE PatchId=%s",
                        PatchId)

    row = InstDB.Cur.fetchone ()
    if not row:
        InstDB.error (req, "Specified patch file not found")
        return

    # Only allow list if patch is active, user is admin or user owns imported file
    if row["State"] != "Active":
        if not InstDB.User or ("Admin" not in InstDB.User["Flags"] and \
              (row["State"] != 'Imported' or InstDB.User["UserId"] != row["UserId"])):
            InstDB.error (req, "Permission denied to specified patch file")
            return            

    itemFields = InstHelper.list_format
    extraFields = ("State", "PatchId")

    # Create SQL query
    sel = InstHelper.sql_select (itemFields + extraFields)

    # Make sure order by field is valid
    if OrderBy and not InstHelper.field_searchable (OrderBy): OrderBy = None
    if not OrderBy: OrderBy = "InstName"

    sel.orderby = InstHelper.sql_order_field (OrderBy, OrderDir)
    sel.limit = "%d,%d" % (Offset, Count)
    sel.where.insert (0, "PatchInfo.PatchId=%d" % PatchId)
    sel.calc_found_rows = True

    InstDB.Cur.execute (sel.query ())
    rows = InstDB.Cur.fetchall ()

    # Get total number of matching items
    InstDB.Cur.execute ("SELECT FOUND_ROWS() AS Total")
    row = InstDB.Cur.fetchone ()
    total = int (row["Total"])

    # Re URL-ify the input fields
    pager_urlstr = urlencode ((("Action", "list"), ("Type", "Inst"),
                               ("OrderBy", OrderBy), ("OrderDir", OrderDir),
                               ("Count", Count), ("PatchId", PatchId)))
    pager_urlstr = cgi.escape (pager_urlstr)

    # Display the pager bar
    InstDB.show_pager (req, pager_urlstr, Offset, Count, total)

    # Create clickable titles that sort output
    urlstr = urlencode ((("Action", "list"), ("Type", "Inst"),
                         ("Count", Count), ("PatchId", PatchId)))
    urlstr = cgi.escape (urlstr)

    sortList = []
    for field in itemFields:
        if InstHelper.field_searchable (field):
            sortList += ((field, InstHelper.field_title (field)),)
        else: sortList += (InstHelper.field_title (field),)

    titles = sort_titles (sortList, urlstr, OrderBy, OrderDir, InstHelper)

    table = InstDB.tabular (titles)
    table.tableattr = 'width="100%"'

    for row in rows:
        req.write (table.addrow (InstHelper.html_vals (itemFields, row)))

    req.write (table.end ())

    # Display the pager bar
    InstDB.show_pager (req, pager_urlstr, Offset, Count, total)


def list_samples (req, fields):
    OrderBy = fields.get ("OrderBy", None)

    OrderDir = int (fields.get ("OrderDir", 0))
    if OrderDir < 0 or OrderDir > 2: OrderDir = 0

    Offset = int (fields.get ("Offset", 0))
    Count = int (fields.get ("Count", 50))
    PatchId = int (fields.get ("PatchId", 0))

    if Count <= 5: Count = 5
    elif Count > InstDB.MaxListItems: Count = InstDB.MaxListItems

    InstDB.Cur.execute ("SELECT State, UserId FROM PatchInfo WHERE PatchId=%s",
                        PatchId)

    row = InstDB.Cur.fetchone ()
    if not row:
        InstDB.error (req, "Specified patch file not found")
        return

    # Only allow list if patch is active, user is admin or user owns imported file
    if row["State"] != "Active":
        if not InstDB.User or ("Admin" not in InstDB.User["Flags"] and \
              (row["State"] != 'Imported' or InstDB.User["UserId"] != row["UserId"])):
            InstDB.error (req, "Permission denied to specified patch file")
            return            

    itemFields = SampleHelper.list_format
    extraFields = ("State",)
    if PatchId: extraFields += ("PatchId",)

    # Create SQL query
    sel = SampleHelper.sql_select (itemFields + extraFields)

    # Make sure order by field is valid
    if OrderBy and not SampleHelper.field_searchable (OrderBy): OrderBy = None
    if not OrderBy: OrderBy = "SampleName"

    sel.orderby = SampleHelper.sql_order_field (OrderBy, OrderDir)
    sel.limit = "%d,%d" % (Offset, Count)
    if PatchId: sel.where.insert (0, "PatchInfo.PatchId=%d" % PatchId)
    sel.calc_found_rows = True

    InstDB.Cur.execute (sel.query ())
    rows = InstDB.Cur.fetchall ()

    # Get total number of matching items
    InstDB.Cur.execute ("SELECT FOUND_ROWS() AS Total");
    row = InstDB.Cur.fetchone ()
    total = int (row["Total"])

    # Re URL-ify the input fields
    pager_urlstr = urlencode ((("Action", "list"), ("Type", "Sample"),
                               ("OrderBy", OrderBy), ("OrderDir", OrderDir),
                               ("Count", Count), ("PatchId", PatchId)))
    pager_urlstr = cgi.escape (pager_urlstr)

    # Display the pager bar
    InstDB.show_pager (req, pager_urlstr, Offset, Count, total)

    # Create clickable titles that sort output
    urlstr = urlencode ((("Action", "list"), ("Type", "Sample"),
                         ("Count", Count), ("PatchId", PatchId)))
    urlstr = cgi.escape (urlstr)

    sortList = []
    for field in itemFields:
        if SampleHelper.field_searchable (field):
            sortList += ((field, SampleHelper.field_title (field)),)
        else: sortList += (SampleHelper.field_title (field),)

    titles = sort_titles (sortList, urlstr, OrderBy, OrderDir, SampleHelper)

    table = InstDB.tabular (titles)
    table.tableattr = 'width="100%"'

    for row in rows:
        req.write (table.addrow (SampleHelper.html_vals (itemFields, row)))

    req.write (table.end ())

    # Display the pager bar
    InstDB.show_pager (req, pager_urlstr, Offset, Count, total)


def sort_titles (titles, urlstr, orderby, orderdir, helper):
    ret = ()
    link = '<a href="patches.py?%s&amp;OrderBy=%s">%s</a>\n'

    for title in titles:
        if type (title) == str: ret += (title,)
        elif orderby != title[0]: ret += (link % (urlstr, title[0], title[1]),)
        else:
            deforder = helper.field_searchable (title[0])
            if orderdir == 0: orderdir = deforder

            # Currently default order? Clicking will alternate it.
            if orderdir == deforder:
                if orderdir == 2: altorder = 1
                else: altorder = 2

                dirstr = "&amp;OrderDir=%d" % altorder
            else: dirstr = ""

            if orderdir == 2:
              arrow = "down"
              alt = "Descending"
            else:
              arrow = "up"
              alt = "Ascending"

            ret += ('[%s]' % title[1]
                    + '<a href="patches.py?%s&amp;OrderBy=%s%s">'
                    % (urlstr, title[0], dirstr)
                    + '<img src="images/arrow_%s.png" alt="%s"' % (arrow, alt)
                    + ' width=16 height=14 border=0></a>',)

    return ret
